contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts);
   }
 
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplySpecificVintages(
+    uint256 amount,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 remainingAmountToFill = amount;
+    uint256 countOfListedRemovals = _removal.numberOfTokensOwnedByAddress({
+      account: address(this)
+    });
+    ids = new uint256[](countOfListedRemovals);
+    amounts = new uint256[](countOfListedRemovals);
+    suppliers = new address[](countOfListedRemovals);
+    countOfRemovalsAllocated = 0;
+    for (uint256 i = 0; i < countOfListedRemovals; ++i) {
+      uint256 removalId = _listedSupply[_currentSupplierAddress]
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any removals from the vintages requested
+      // at that point we should simply move on to the next supplier
+      if (removalId == 0) {
+        _incrementCurrentSupplierAddress();
```

Comment:
Won't this push the pointer of the round robin order to the next supplier? I don't think purchasing from specific vintages (and jumping outside of the queue as ordered) should modify the state of the queue. Can we store this value and reset it at the end of the function (unless we otherwise detect that we legitimately need to increment the supplier)?

Reply:
yeah that's a good idea

Reply:
i'm going to do this in a slightly different way... if all we do is keep track of the `_currentSupplierAddress` before the fulfillment algorithm so that we can reset it later, this supplier may be removed during fulfillment as well as any number of other intervening suppliers (but not necessarily all of them as this fulfillment pulls only from supplier with certain vintages).  I think it's easier to reason about it if we have a local `localCurrentSupplier` that starts as the global `_currentSupplierAddress` and then serves the same role.  the only tricky case here is that if we remove the `localCurrentSupplier` while removing the removal, we have to be responsible for making sure we move on to the next supplier locally by checking if the supplier we were on is now empty.  the `removeActiveRemoval` function does what we need it to in terms of keeping the global data structure and `_currentSupplierAddress` consistent which is great.